File: 2024 December/Longest Consecutive Subsequence/main.cpp
// Link: https://www.geeksforgeeks.org/problems/longest-consecutive-subsequence2449/1

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int longestConsecutive(vector<int> &arr)
    {
        unordered_map<int, int> mp;
        int ans = 0;
        for (int i = 0; i < arr.size(); i++)
        {
            mp[arr[i]]++;
        }
        for (int i = 0; i < arr.size(); i++)
        {
            int x = arr[i];
            if (mp[x - 1] == 0)
            {
                while (mp[x] != 0)
                    x++;

                ans = max(ans, x - arr[i]);
            }
        }
        return ans;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {2, 6, 1, 9, 4, 5, 3};
    cout << object.longestConsecutive(arr) << endl;

    arr = {1, 9, 3, 10, 4, 20, 2};
    cout << object.longestConsecutive(arr) << endl;

    arr = {15, 13, 12, 14, 11, 10, 9};
    cout << object.longestConsecutive(arr) << endl;
    return 0;
}

File: 2024 December/Longest Consecutive Subsequence/main.py
# Link: https://www.geeksforgeeks.org/problems/longest-consecutive-subsequence2449/1


class Solution:
    def longestConsecutive(self, arr):
        mp = {}
        ans = 0
        for num in arr:
            mp[num] = mp.get(num, 0) + 1
        for num in arr:
            if mp.get(num - 1, 0) == 0:
                x = num
                while mp.get(x, 0) != 0:
                    x += 1
                ans = max(ans, x - num)
        return ans


if __name__ == "__main__":
    object = Solution()
    arr = [2, 6, 1, 9, 4, 5, 3]
    print(object.longestConsecutive(arr))

    arr = [1, 9, 3, 10, 4, 20, 2]
    print(object.longestConsecutive(arr))

    arr = [15, 13, 12, 14, 11, 10, 9]
    print(object.longestConsecutive(arr))
